aten/src/ATen/native/Batching.cpp

```diff
@@ -20,6 +21,73 @@ static bool has_level(const Tensor& self, int64_t level) {
   return it != bdims.end();
 }
 
+// Returns a Tensor with batch dim with level `level` turned into a regular dimension,
+// as well as a logical dim index of where said dimension is in the returned tensor.
+//
+// The reason why we want to return the index of where said dimension is in the returned
+// tensor is because we want to keep track of which dimension used to be the batch
+// dimension so that we can move it to the correct logical dimension specified by
+// `out_dims` in vmap. For example, if we had
+// >>> x = torch.randn(2, 3, 5)
+// >>> vmap(lambda x: x, in_dims=0, out_dims=1)(x)
+// then right when we are about to exit the vmap block, x is a BatchedTensor with a
+// batch dimension at (physical) index 0. Note that the batch dimension doesn't
+// always have to exist at (physical) index 0. When we undo the batch dimension,
+// we want to move it to dimension 1 (as specified by out_dims). So we return the
+// index at which the batch dim appears so that we can move it to the correct place.
+// later down the line.
+static std::pair<Tensor,int64_t> remove_existing_batch_dim(const Tensor& self, int64_t level) {
+  const auto* batched = maybeGetBatched(self);
+  TORCH_INTERNAL_ASSERT(batched != nullptr);
+  auto bdims = batched->bdims();
+  if (bdims.size() == 1) {
+    TORCH_INTERNAL_ASSERT(bdims[0].level() == level);
+    return std::make_pair(batched->value(), bdims[0].dim());
+  }
+  BatchDims new_bdims;
+  int64_t newly_exposed_physical_dim;
+  new_bdims.reserve(bdims.size() - 1);
+  for (const auto& bdim : bdims) {
+    if (bdim.level() == level) {
+      newly_exposed_physical_dim = bdim.dim();
+    } else {
+      new_bdims.push_back(bdim);
+    }
+  }
+  int64_t num_batch_dims_before_newly_exposed_physical_dim = std::count_if(
+      new_bdims.begin(), new_bdims.end(),
+      [&](const BatchDim& bdim) {
+        return bdim.dim() < newly_exposed_physical_dim;
+      });
+  int64_t newly_exposed_logical_dim =
+      newly_exposed_physical_dim - num_batch_dims_before_newly_exposed_physical_dim;
+  auto result_tensor = makeBatched(batched->value(), std::move(new_bdims));
+  return std::make_pair(std::move(result_tensor), newly_exposed_logical_dim);
+}
+
+// Poor man's version of np.moveaxis. Moves the dimension at `dst` to `src`
+// while preserving the order of other existing dimensions.
+// We should probably add np.moveaxis (it is more general) to PyTorch. (#36048)
+// When we do, replace the following with it.
+static Tensor movedim(const Tensor& self, int64_t src, int64_t dst) {
+  auto logical_dim = self.dim();
+  src = maybe_wrap_dim(src, logical_dim);
+  dst = maybe_wrap_dim(dst, logical_dim);
+  if (src == dst) {
+    return self;
+  }
+  VmapDimVector permutation;
+  permutation.reserve(logical_dim);
+  for (int64_t dim = 0; dim < logical_dim; dim++) {
+    if (dim == src) {
+      continue;
+    }
+    permutation.push_back(dim);
+  }
+  permutation.insert(permutation.begin() + dst, src);
```

Comment:
O(n) police! (Should be fine)

Reply:
If you saw the implementation of np.moveaxis... there would be an O(n^2) police sounding alarms. Otherwise I would have just ported np.moveaxis in PyTorch and used it here